crypto/fipsmodule/ec/ec.c

```diff
@@ -391,46 +407,103 @@ EC_GROUP *EC_GROUP_new_by_curve_name(int nid) {
   }
 }
 
+EC_GROUP *EC_GROUP_new_by_curve_name_mutable(int nid) {
+  EC_GROUP *ret = NULL;
+  switch (nid) {
+    case NID_secp224r1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_p224(), sizeof(EC_GROUP));
+      break;
+    case NID_X9_62_prime256v1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_p256(), sizeof(EC_GROUP));
+      break;
+    case NID_secp384r1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_p384(), sizeof(EC_GROUP));
+      break;
+    case NID_secp521r1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_p521(), sizeof(EC_GROUP));
+      break;
+    case NID_secp256k1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_secp256k1(), sizeof(EC_GROUP));
+      break;
+    default:
+      OPENSSL_PUT_ERROR(EC, EC_R_UNKNOWN_GROUP);
+      return NULL;
+  }
+  if (ret == NULL) {
+    return NULL;
+  }
+  ret->mutable_ec_group = 1;
+  return ret;
+}
+
 void EC_GROUP_free(EC_GROUP *group) {
-  if (group == NULL ||
-      // Built-in curves are static.
-      group->curve_name != NID_undef ||
-      !CRYPTO_refcount_dec_and_test_zero(&group->references)) {
+  if (group == NULL) {
     return;
   }
 
+  if (!group->mutable_ec_group) {
+    if (group->curve_name != NID_undef ||
```

Comment:
are we sure that custom curves will always have `curve_name == NID_undef`? because we currently treat all custom curves as "immutable", if the caller sets some other NID on a custom curve, we'd decrement the refcount and return instead of free'ing. is that what we want?

Reply:
From what I've observed for our `EC_GROUP` implementations it does seem like this is the main differentiation between custom and named curves. We don't expose any methods to set an NID for custom curves, although that may change if we are ever forced to implement `EC_GROUP_set_curve_name`. 

It does make me feel a bit more at ease, when looking at [OpenSSL's logic ](https://github.com/openssl/openssl/blob/2f362e99a1178263c7102474f0190836166f416d/crypto/ec/ec_lib.c#L483-L490)for the function though.
```
void EC_GROUP_set_curve_name(EC_GROUP *group, int nid)
{
    group->curve_name = nid;
    group->asn1_flag =
        (nid != NID_undef)
        ? OPENSSL_EC_NAMED_CURVE
        : OPENSSL_EC_EXPLICIT_CURVE;
}
```
Only named curves are expected to have NIDs, explicit/custom curves follow the same logic of having `NID_undef` in OpenSSL as well.